src/main/java/org/gaul/s3proxy/AliasBlobStore.java

```diff
@@ -0,0 +1,297 @@
+import com.google.common.collect.ImmutableList;
+
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.domain.Blob;
+import org.jclouds.blobstore.domain.BlobMetadata;
+import org.jclouds.blobstore.domain.ContainerAccess;
+import org.jclouds.blobstore.domain.MultipartPart;
+import org.jclouds.blobstore.domain.MultipartUpload;
+import org.jclouds.blobstore.domain.MutableStorageMetadata;
+import org.jclouds.blobstore.domain.PageSet;
+import org.jclouds.blobstore.domain.StorageMetadata;
+import org.jclouds.blobstore.domain.internal.MutableStorageMetadataImpl;
+import org.jclouds.blobstore.domain.internal.PageSetImpl;
+import org.jclouds.blobstore.options.CopyOptions;
+import org.jclouds.blobstore.options.CreateContainerOptions;
+import org.jclouds.blobstore.options.GetOptions;
+import org.jclouds.blobstore.options.ListContainerOptions;
+import org.jclouds.blobstore.options.PutOptions;
+import org.jclouds.blobstore.util.ForwardingBlobStore;
+import org.jclouds.domain.Location;
+import org.jclouds.io.Payload;
+
+/**
+ * This class implements a middleware to alias buckets to a different name.
+ * The aliases are configured as:
+ *   s3proxy.alias-blobstore.&lt;alias name&gt; = &lt;backend bucket&gt;
+ *
+ * The aliases appear in bucket listings if the configured
+ * backend buckets are present. Requests for all other buckets are unaffected.
+ */
+public final class AliasBlobStore extends ForwardingBlobStore {
+    private static final Pattern ALIAS_RE = Pattern.compile(
+            S3ProxyConstants.PROPERTY_ALIAS_BLOBSTORE + "\\.(?<bucket>.*)$");
+
+    private final BiMap<String, String> aliases;
+
+    private AliasBlobStore(BlobStore delegate,
+                           BiMap<String, String> aliases) {
+        super(delegate);
+        this.aliases = requireNonNull(aliases);
+    }
+
+    static BlobStore newAliasBlobStore(BlobStore delegate,
+                                       BiMap<String, String> aliases) {
+        return new AliasBlobStore(delegate, aliases);
+    }
+
+    private MultipartUpload getDelegateMpu(MultipartUpload mpu) {
+        return MultipartUpload.create(
+                getContainer(mpu.containerName()),
+                mpu.blobName(),
+                mpu.id(),
+                mpu.blobMetadata(),
+                mpu.putOptions());
+    }
+
+    public static ImmutableBiMap<String, String> parseAliases(
+            Properties properties) {
+        ImmutableBiMap.Builder<String, String> builder =
+                new ImmutableBiMap.Builder<>();
+        Set<String> backendBuckets = new HashSet<>();
+        for (String key : properties.stringPropertyNames()) {
+            Matcher matcher = ALIAS_RE.matcher(key);
+            if (matcher.matches()) {
+                String virtualBucket = matcher.group("bucket");
+                String backendBucket = properties.getProperty(key);
+                checkArgument(
+                        !backendBuckets.contains(properties.getProperty(key)),
```

Comment:
Triplicated call to `properties.getProperty`.

Reply:
Done